File: sdks/elixir/lib/hookrelay/webhook_verification.ex
defmodule HookRelay.WebhookVerification do
  @moduledoc """
  Webhook signature verification for HookRelay.

  Supports both simple HMAC-SHA256 verification and Standard Webhooks
  (Svix-compatible) verification with timestamp tolerance.

  Also supports Svix headers (svix-id, svix-signature, svix-timestamp) as fallback.

  ## Example

      # Verify from request headers
      result = HookRelay.WebhookVerification.verify_webhook_from_headers(
        payload: body,
        headers: conn.req_headers |> Map.new(),
        secret: "whsec_..."
      )

      case result do
        {:ok, payload} -> # Valid webhook
        {:error, reason} -> # Invalid
      end
  """

  @default_tolerance_secs 300

  @doc """
  Verify a webhook signature using HMAC-SHA256 (legacy format).
  """
  @spec verify_signature(String.t(), String.t(), String.t()) :: boolean()
  def verify_signature(payload, signature, secret)
      when is_binary(payload) and is_binary(signature) and is_binary(secret) do
    if payload == "" or signature == "" or secret == "" do
      false
    else
      expected_hex =
        if String.starts_with?(signature, "sha256="),
          do: String.slice(signature, 7..-1//1),
          else: signature

      computed = :crypto.mac(:hmac, :sha256, secret, payload) |> Base.encode16(case: :lower)
      Plug.Crypto.secure_compare(computed, expected_hex)
    end
  end

  def verify_signature(_, _, _), do: false

  @doc """
  Verify a webhook using Standard Webhooks headers (Svix-compatible).

  Returns `{:ok, parsed_payload}` or `{:error, reason}`.
  """
  @spec verify_webhook(keyword()) :: {:ok, map() | String.t()} | {:error, String.t()}
  def verify_webhook(opts) do
    payload = Keyword.fetch!(opts, :payload)
    msg_id = Keyword.get(opts, :msg_id)
    timestamp = Keyword.get(opts, :timestamp)
    signature_header = Keyword.get(opts, :signature_header)
    secret = Keyword.fetch!(opts, :secret)
    tolerance_secs = Keyword.get(opts, :tolerance_secs, @default_tolerance_secs)

    cond do
      is_nil(msg_id) or msg_id == "" ->
        {:error, "Missing webhook-id header"}

      is_nil(timestamp) or timestamp == "" ->
        {:error, "Missing webhook-timestamp header"}

      is_nil(signature_header) or signature_header == "" ->
        {:error, "Missing webhook-signature header"}

      is_nil(payload) or payload == "" ->
        {:error, "Missing request body"}

      true ->
        verify_webhook_impl(payload, msg_id, timestamp, signature_header, secret, tolerance_secs)
    end
  end

  @doc """
  Verify a webhook from headers with automatic header detection.

  Supports both Standard Webhooks headers (webhook-id, webhook-signature, webhook-timestamp)
  and Svix headers (svix-id, svix-signature, svix-timestamp) as fallback.
  """
  @spec verify_webhook_from_headers(keyword()) :: {:ok, map() | String.t()} | {:error, String.t()}
  def verify_webhook_from_headers(opts) do
    headers = Keyword.fetch!(opts, :headers)
    normalized = for {k, v} <- headers, into: %{}, do: {String.downcase(to_string(k)), v}

    msg_id = normalized["webhook-id"]
    timestamp = normalized["webhook-timestamp"]
    signature_header = normalized["webhook-signature"]

    {msg_id, timestamp, signature_header} =
      if is_nil(msg_id) or is_nil(timestamp) or is_nil(signature_header) do
        {
          msg_id || normalized["svix-id"],
          timestamp || normalized["svix-timestamp"],
          signature_header || normalized["svix-signature"]
        }
      else
        {msg_id, timestamp, signature_header}
      end

    verify_webhook(
      payload: opts[:payload],
      msg_id: msg_id,
      timestamp: timestamp,
      signature_header: signature_header,
      secret: opts[:secret],
      tolerance_secs: opts[:tolerance_secs]
    )
  end

  # Private implementation

  defp verify_webhook_impl(payload, msg_id, timestamp, signature_header, secret, tolerance_secs) do
    case Integer.parse(timestamp) do
      {ts, ""} ->
        now = System.system_time(:second)

        if now - ts > tolerance_secs do
          {:error, "Message timestamp too old"}
        else
          if ts > now + tolerance_secs do
            {:error, "Message timestamp too new"}
          else
            do_verify_signature(payload, msg_id, timestamp, signature_header, secret)
          end
        end

      _ ->
        {:error, "Invalid webhook timestamp"}
    end
  end

  defp do_verify_signature(payload, msg_id, timestamp, signature_header, secret) do
    signed_content = "#{msg_id}.#{timestamp}.#{payload}"
    secret_bytes = decode_secret(secret)

    expected_sig =
      :crypto.mac(:hmac, :sha256, secret_bytes, signed_content)
      |> Base.encode64()

    expected_full = "v1,#{expected_sig}"

    signatures = String.split(signature_header, " ")

    verified =
      Enum.any?(signatures, fn sig ->
        trimmed = String.trim(sig)
        String.starts_with?(trimmed, "v1,") and Plug.Crypto.secure_compare(trimmed, expected_full)
      end)

    if verified do
      case Jason.decode(payload) do
        {:ok, parsed} -> {:ok, parsed}
        {:error, _} -> {:ok, payload}
      end
    else
      {:error, "Invalid webhook signature"}
    end
  end

  defp decode_secret(secret) do
    stripped =
      if String.starts_with?(secret, "whsec_"),
        do: String.slice(secret, 6..-1//1),
        else: secret

    # Add padding in case secret is unpadded base64
    case Base.decode64(stripped <> "==") do
      {:ok, decoded} -> decoded
      :error -> secret
    end
  end
end

File: sdks/java/src/main/java/com/hookrelay/WebhookVerification.java
package com.hookrelay;

import com.google.gson.Gson;
import com.google.gson.JsonObject;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Base64;
import java.util.Map;

/**
 * Webhook signature verification for HookRelay.
 *
 * <p>Supports both simple HMAC-SHA256 verification and Standard Webhooks
 * (Svix-compatible) verification with timestamp tolerance.</p>
 */
public class WebhookVerification {

    private static final int DEFAULT_TOLERANCE_SECS = 300;
    private static final Gson GSON = new Gson();

    /**
     * Verify a webhook signature using HMAC-SHA256.
     *
     * @param payload   The raw request body
     * @param signature The signature from the X-Hookrelay-Signature header
     * @param secret    The endpoint's signing secret (starts with "whsec_")
     * @return true if the signature is valid
     */
    public static boolean verifySignature(String payload, String signature, String secret) {
        if (payload == null || payload.isEmpty()) return false;
        if (signature == null || signature.isEmpty()) return false;
        if (secret == null || secret.isEmpty()) return false;

        String expectedHex = signature.startsWith("sha256=") ? signature.substring(7) : signature;

        try {
            String computed = hmacSha256Hex(secret, payload);
            return constantTimeEquals(computed, expectedHex);
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Verify a webhook using Standard Webheaders headers (Svix-compatible).
     *
     * @param payload         The raw request body
     * @param msgId           The webhook-id header
     * @param timestamp       The webhook-timestamp header
     * @param signatureHeader The webhook-signature header
     * @param secret          The endpoint's signing secret
     * @return VerificationResult with valid flag and parsed payload
     */
    public static VerificationResult verifyWebhook(String payload, String msgId,
                                                    String timestamp, String signatureHeader,
                                                    String secret) {
        return verifyWebhook(payload, msgId, timestamp, signatureHeader, secret, DEFAULT_TOLERANCE_SECS);
    }

    /**
     * Verify a webhook using Standard Webheaders headers with custom tolerance.
     */
    public static VerificationResult verifyWebhook(String payload, String msgId,
                                                    String timestamp, String signatureHeader,
                                                    String secret, int toleranceSecs) {
        if (msgId == null || msgId.isEmpty()) {
            return VerificationResult.invalid("Missing webhook-id header");
        }
        if (timestamp == null || timestamp.isEmpty()) {
            return VerificationResult.invalid("Missing webhook-timestamp header");
        }
        if (signatureHeader == null || signatureHeader.isEmpty()) {
            return VerificationResult.invalid("Missing webhook-signature header");
        }
        if (payload == null || payload.isEmpty()) {
            return VerificationResult.invalid("Missing request body");
        }

        // Validate timestamp
        long ts;
        try {
            ts = Long.parseLong(timestamp);
        } catch (NumberFormatException e) {
            return VerificationResult.invalid("Invalid webhook timestamp");
        }

        long now = System.currentTimeMillis() / 1000;
        long age = Math.abs(now - ts);

        if (age > toleranceSecs) {
            return VerificationResult.invalid(
                String.format("Webhook timestamp expired: %ds old (tolerance: %ds)", age, toleranceSecs));
        }

        // Compute expected signature
        String signedContent = msgId + "." + timestamp + "." + payload;
        byte[] secretBytes = decodeSecret(secret);

        try {
            String expectedSig = Base64.getEncoder().encodeToString(hmacSha256(secretBytes, signedContent));
            String expectedFull = "v1," + expectedSig;

            // Check each signature in the header (space-separated)
            String[] signatures = signatureHeader.split(" ");
            boolean verified = false;

            for (String sig : signatures) {
                String trimmed = sig.trim();
                if (!trimmed.startsWith("v1,")) continue;

                if (constantTimeEquals(trimmed, expectedFull)) {
                    verified = true;
                    break;
                }
            }

            if (!verified) {
                return VerificationResult.invalid("Invalid webhook signature");
            }
        } catch (Exception e) {
            return VerificationResult.invalid("Signature computation failed");
        }

        // Parse payload
        try {
            JsonObject parsed = GSON.fromJson(payload, JsonObject.class);
            return VerificationResult.valid(parsed);
        } catch (Exception e) {
            return VerificationResult.valid(payload);
        }
    }

    // ==================== Helpers ====================

    private static String hmacSha256Hex(String key, String data) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec(key.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
        byte[] hash = mac.doFinal(data.getBytes(StandardCharsets.UTF_8));
        return bytesToHex(hash);
    }

    private static byte[] hmacSha256(byte[] key, String data) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec(key, "HmacSHA256"));
        return mac.doFinal(data.getBytes(StandardCharsets.UTF_8));
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            sb.append(String.format("%02x", b & 0xff));
        }
        return sb.toString();
    }

    private static boolean constantTimeEquals(String a, String b) {
        if (a == null || b == null) return false;
        return MessageDigest.isEqual(
            a.getBytes(StandardCharsets.UTF_8),
            b.getBytes(StandardCharsets.UTF_8)
        );
    }

    private static byte[] decodeSecret(String secret) {
        String stripped = secret.startsWith("whsec_") ? secret.substring(6) : secret;
        try {
            return Base64.getDecoder().decode(stripped);
        } catch (IllegalArgumentException e) {
            return secret.getBytes(StandardCharsets.UTF_8);
        }
    }

    /**
     * Result of webhook verification.
     */
    public static class VerificationResult {
        private final boolean valid;
        private final Object payload;
        private final String error;

        private VerificationResult(boolean valid, Object payload, String error) {
            this.valid = valid;
            this.payload = payload;
            this.error = error;
        }

        static VerificationResult valid(Object payload) {
            return new VerificationResult(true, payload, null);
        }

        static VerificationResult invalid(String error) {
            return new VerificationResult(false, null, error);
        }

        public boolean isValid() { return valid; }
        public Object getPayload() { return payload; }
        public String getError() { return error; }
    }
}
